tests/phpunit/SQLStore/QueryEngine/QueryEngineTest.php

```diff
@@ -54,6 +55,61 @@ public function testGetQueryResultForDebugQueryMode() {
 			->disableOriginalConstructor()
 			->getMock();
 
+		$queryBuilder = $this->getMockBuilder( SelectQueryBuilder::class )
+			->disableOriginalConstructor()
+			->getMock();
+
+		$queryBuilder->expects( $this->any() )
+			->method( 'select' )
+			->willReturn( $queryBuilder );
+
+		$queryBuilder->expects( $this->any() )
+			->method( 'from' )
+			->willReturn( $queryBuilder );
+
+		$queryBuilder->expects( $this->any() )
+			->method( 'where' )
+			->willReturn( $queryBuilder );
+
+		$queryBuilder->expects( $this->any() )
+			->method( 'distinct' )
+			->willReturn( $queryBuilder );
+
+		$queryBuilder->expects( $this->any() )
+			->method( 'caller' )
+			->willReturn( $queryBuilder );
+
+		// Mock getSQL() for debug mode
+		$queryBuilder->expects( $this->any() )
+			->method( 'getSQL' )
+			->willReturn( 'SELECT * FROM test' );
+
+		$connection->expects( $this->any() )
+			->method( 'newSelectQueryBuilder' )
+			->willReturn( $queryBuilder );
+
+		$connection->expects( $this->any() )
+			->method( 'applySqlOptions' )
+			->willReturn( $queryBuilder );
+
+		// Mock isType() for EXPLAIN query formatting
+		$connection->expects( $this->any() )
+			->method( 'isType' )
+			->willReturn( false );
+
+		// Mock query() for EXPLAIN
+		$resultWrapper = $this->getMockBuilder( '\Wikimedia\Rdbms\IResultWrapper' )
+			->disableOriginalConstructor()
+			->getMock();
+
+		$resultWrapper->expects( $this->any() )
+			->method( 'free' )
+			->willReturn( true );
+
+		$connection->expects( $this->any() )
+			->method( 'query' )
+			->willReturn( $resultWrapper );
```

Comment:
Offf. That is a lot of mocking. Room for extraction into private methods, to avoid 100 lines of mocking obscuring the high level happenings in the test.

---

Don't do

```
		$queryBuilder = $this->getMockBuilder( SelectQueryBuilder::class )
			->disableOriginalConstructor()
			->getMock();
```

Do `$queryBuilder = $this->createMock( SelectQueryBuilder::class );`

---

Also see https://www.entropywins.wtf/blog/2016/05/13/5-ways-to-write-better-mocks/

Might be difficult to apply here due to tech debt in MW/SMW

Reply:
:100: